Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt

```diff
@@ -0,0 +1,249 @@
+                for (rollbackItem: RollbackItem in rollbackItems) rollbackItem.invoke()
+            } catch (rollbackException: Exception) {
+                Timber.tag(TAG).e(rollbackException, "RiskLevelRepository rollback failed.")
+            }
+
+            throw error
+        }
+    }
+
+    override fun calculationNotPossibleBecauseOfOutdatedResults(): Boolean {
+        // if the last calculation is longer in the past as the defined threshold we return the stale state
+        val timeSinceLastDiagnosisKeyFetchFromServer =
+            TimeVariables.getTimeSinceLastDiagnosisKeyFetchFromServer()
+                ?: throw RiskLevelCalculationException(
+                    IllegalArgumentException(
+                        "Time since last exposure calculation is null"
+                    )
+                )
+        /** we only return outdated risk level if the threshold is reached AND the active tracing time is above the
+        defined threshold because [UNKNOWN_RISK_INITIAL] overrules [UNKNOWN_RISK_OUTDATED_RESULTS] */
+        return timeSinceLastDiagnosisKeyFetchFromServer.millisecondsToHours() >
+            TimeVariables.getMaxStaleExposureRiskRange() && isActiveTracingTimeAboveThreshold()
+    }
+
+    override fun calculationNotPossibleBecauseNoKeys() =
+        (TimeVariables.getLastTimeDiagnosisKeysFromServerFetch() == null).also {
+            if (it) {
+                Timber.tag(TAG)
+                    .v("No last time diagnosis keys from server fetch timestamp was found")
+            }
+        }
+
+    override suspend fun calculationNotPossibleBecauseTracingIsOff() =
+        // this applies if tracing is not activated
+        !enfClient.isTracingEnabled.first()
+
+    override suspend fun isIncreasedRisk(): Boolean {
+        val lastExposureSummary = getNewExposureSummary()
+        val appConfiguration = appConfigProvider.getAppConfig()
+        Timber.tag(TAG).v("Retrieved configuration from backend")
+        // custom attenuation parameters to weight the attenuation
+        // values provided by the Google API
+        val attenuationParameters = appConfiguration.attenuationDuration
+        // these are the defined risk classes. They will divide the calculated
+        // risk score into the low and increased risk
+        val riskScoreClassification = appConfiguration.riskScoreClasses
+
+        // calculate the risk score based on the values collected by the Google EN API and
+        // the backend configuration
+        val riskScore = calculateRiskScore(
+            attenuationParameters,
+            lastExposureSummary
+        ).also {
+            Timber.tag(TAG).v("Calculated risk with the given config: $it")
+        }
+
+        // get the high risk score class
+        val highRiskScoreClass =
+            riskScoreClassification.riskClassesList.find { it.label == "HIGH" }
+                ?: throw RiskLevelCalculationException(IllegalStateException("No high risk score class found"))
```

Comment:
is there a more robust way to handle this?

Reply:
What do you mean by 'robust'? An exception has to be thrown in order to abort the task. All exceptions are caught by the task controller.

Reply:
the exception is fine, i was just thinking about the whole process of resolving the highest risk class by comparing a list of instance properties to a string.

Reply:
kind of seems like this might fail if the classes are changed and we forget to update here, there is no real hard coupling that enforces compile time correctness

Reply:
i'll resolve this as we discussed a fix for it internally